### PR TITLE
Firmware inventory datasource

### PR DESCRIPTION
**Pull request by Sakshi-dell**
Firmware inventory datasource added
image

### CODE REVIEW

**Review by Krishnan-Priyanshu — APPROVED** [review]

File: redfish/provider/data_source_redfish_firmware_inventory.go [hunk 1]
```diff
@@ -1,173 +1,150 @@
...
+
+func readRedfishFirmwareInventory(service *gofish.Service) (*models.FirmwareInventory, error) {
+	updateService, err := service.UpdateService()
+	if err != nil {
+		return nil, fmt.Errorf("error fetching UpdateService collection: %s", err.Error())
+	}
+
+	fwInventories, err := updateService.FirmwareInventories()
+	if err != nil {
+		return nil, fmt.Errorf("error fetching Firmware Inventory: %s", err)
```
Comment (rounak-adhikary) [inline redfish/provider/data_source_redfish_firmware_inventory.go:138]: This wont give correct error message. U need to error wrap.
```suggestion
		return nil, fmt.Errorf("error fetching Firmware Inventory: %w", err)
```

File: redfish/provider/data_source_redfish_firmware_inventory.go [hunk 2]
```diff
@@ -1,173 +1,150 @@
...
+			inventoryItemList = append(inventoryItemList, inv)
+		}
+	}
+	return inventoryItemList
+}
+
+func readRedfishFirmwareInventory(service *gofish.Service) (*models.FirmwareInventory, error) {
+	updateService, err := service.UpdateService()
+	if err != nil {
+		return nil, fmt.Errorf("error fetching UpdateService collection: %s", err.Error())
```
Comment (rounak-adhikary) [inline redfish/provider/data_source_redfish_firmware_inventory.go:133]: Error wrapping would be preferred here too
```suggestion
		return nil, fmt.Errorf("error fetching UpdateService collection: %w", err)
```

File: redfish/provider/data_source_redfish_firmware_inventory.go [hunk 3]
```diff
@@ -1,173 +1,150 @@
...
+
+	inventoryItemList := make([]models.Inventory, 0)
+
+	for _, fwInv := range fwInventories {
+		if strings.HasPrefix(fwInv.Entity.ID, "Installed") {
+			inv.EntityId = types.StringValue(fwInv.Entity.ID)
+			inv.EntityName = types.StringValue(fwInv.Entity.Name)
+			inv.Version = types.StringValue(fwInv.Version)
+			inventoryItemList = append(inventoryItemList, inv)
+		}
```
Comment (rounak-adhikary) [inline redfish/provider/data_source_redfish_firmware_inventory.go:125]: I would suggest de-indentation and restricting the scope of inv
```suggestion
		if !strings.HasPrefix(fwInv.Entity.ID, "Installed") {
			continue
		}
		inventoryItemList = append(inventoryItemList, models.Inventory{
			EntityId: types.StringValue(fwInv.Entity.ID)
			EntityName: types.StringValue(fwInv.Entity.Name)
			Version: types.StringValue(fwInv.Version)
		})
```

**Review by rounak-adhikary — APPROVED** [review]
LGTM

**Review by Krishnan-Priyanshu — APPROVED** [review]

**Pull request merged**